echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/cdevents/CDEventsBuilderService.java

```diff
@@ -0,0 +1,200 @@
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+public class CDEventsBuilderService {
+
+  public CloudEvent createCDEvent(
+      Map<String, Object> preference,
+      String application,
+      Event event,
+      Map<String, String> config,
+      String status,
+      String spinnakerUrl) {
+
+    String configType = Optional.ofNullable(config).map(c -> (String) c.get("type")).orElse(null);
+    String configLink = Optional.ofNullable(config).map(c -> (String) c.get("link")).orElse(null);
+
+    String executionId =
+        Optional.ofNullable(event.content)
+            .map(e -> (Map) e.get("execution"))
+            .map(e -> (String) e.get("id"))
+            .orElse(null);
+
+    String executionUrl =
+        String.format(
+            "%s/#/applications/%s/%s/%s",
+            spinnakerUrl,
+            application,
+            configType == "stage" ? "executions/details" : configLink,
+            executionId);
+
+    String executionName =
+        Optional.ofNullable(event.content)
+            .map(e -> (Map) e.get("execution"))
+            .map(e -> (String) e.get("name"))
+            .orElse(null);
+
+    String cdEventsType =
+        Optional.ofNullable(preference).map(p -> (String) p.get("cdEventsType")).orElse(null);
+
+    CloudEvent ceToSend =
+        buildCloudEventWithCDEventType(
+            cdEventsType, executionId, executionUrl, executionName, spinnakerUrl, status);
+    if (ceToSend == null) {
+      log.error("Failed to created CDEvent with type {} as CloudEvent", cdEventsType);
+      throw new CDEventsException("Failed to created CDEvent as CloudEvent");
+    }
+    return ceToSend;
+  }
+
+  private CloudEvent buildCloudEventWithCDEventType(
+      String cdEventsType,
+      String executionId,
+      String executionUrl,
+      String executionName,
+      String spinnakerUrl,
+      String status) {
+    CloudEvent ceToSend = null;
+    switch (cdEventsType) {
+      case "dev.cdevents.pipelinerun.queued":
+        ceToSend =
+            createPipelineRunQueuedEvent(executionId, executionUrl, executionName, spinnakerUrl);
+        break;
+      case "dev.cdevents.pipelinerun.started":
+        ceToSend =
+            createPipelineRunStartedEvent(executionId, executionUrl, executionName, spinnakerUrl);
+        break;
+      case "dev.cdevents.pipelinerun.finished":
+        ceToSend =
+            createPipelineRunFinishedEvent(
+                executionId, executionUrl, executionName, spinnakerUrl, status);
+        break;
+      case "dev.cdevents.taskrun.started":
+        ceToSend =
+            createTaskRunStartedEvent(executionId, executionUrl, executionName, spinnakerUrl);
+        break;
+      case "dev.cdevents.taskrun.finished":
+        ceToSend =
+            createTaskRunFinishedEvent(
+                executionId, executionUrl, executionName, spinnakerUrl, status);
+        break;
+      default:
+        throw new CDEventsException(
```

Comment:
As per my comment in the PR description, I will update the same PR/create another PR for all other events once this review is closed. 
`Current PR has the implementation to produce [Core CDEvents](https://github.com/cdevents/spec/blob/v0.1.2/core.md) for now, other events will be implemented after this PR review and discussion.`

My question here is does Spinnaker needs all types of events which CDEvents spec provides(https://cdevents.dev/docs/) or needs a subset of event types based on any Spinnaker requirements.
Like I am not sure whether [Source Code Control Events](https://cdevents.dev/docs/source-code-version-control/) are required for Spinnaker to send.

There is no such interface to create CDEvents(might be a good idea to propose this change from SDK itself as that interface/implementations can be used as common), 
I can update the current switch case to a map for flexibility and extensibility purpose.

Reply:
> There is no such interface to create CDEvents(might be a good idea to propose this change from SDK itself as that interface/implementations can be used as common)

What I was meaning though is:

```java
public interface CDEventCreateTask {
    public CloudEvent create(String executionId, String executionUrl, String executionName, String spinnakerUrl, String status);
}
```

```java
public class CDEventCreateTaskRunFinished implements CDEventCreateTask {
// Code here
}
```

Then in this class

```java
private final Map<CDEventTypeEnum, CDEventCreateTask> createTasks = Map.of (
    CDEventTaskRunFinished, new CDEventCreateTaskRunFinished(),
);
```

That way the switch statement goes away and instead becomes something like

```java
CDEventCreateTask createTask = createTasks.get(eventTypeEnum);
if (createTask == null) {
    throw new SomeException("Invalid CDEvent create task");
}
```

Reply:
yes I'm in the same direction, was thinking aloud rather implementing in the Spinnaker is this a good idea to create these interface and classes in the CDEvents Java-SDK itself, 
but I know It will restrict to create CDEvents the way Spinnaker wants with any additional changes.
Will create these classes with little modifications here.
```java
 public interface CDEventCreator {
 //other type of events can have different parameters, so will remove the params
  CloudEvent createCDEvent();
}
```
Implement the classes for each event
```java
public class CDEventPipelineRunQueued implements CDEventCreator {

//each event can have different parameters, can go here
  private String source;
  private String subjectId;
  private String subjectSource;
  private String subjectPipelineName;
  private String subjectUrl;

  public CDEventPipelineRunQueued(String executionId, String executionUrl, String executionName, String spinnakerUrl) {
    this.source = spinnakerUrl;
    this.subjectId = executionId;
    this.subjectSource = spinnakerUrl;
    this.subjectPipelineName = executionName;
    this.subjectUrl = executionUrl;
  }

  @Override
  public CloudEvent createCDEvent() {
PipelineRunQueuedCDEvent cdEvent = new PipelineRunQueuedCDEvent();
}
```
creating a map of different events in this class,
```java
Map<CDEventTypeEnum, CDEventCreateTask> createTasks = Map.of (
      PipelineRunQueuedEnum, new CDEventPipelineRunQueued(executionId, executionUrl, executionName, spinnakerUrl)
    );
```